docs/decisions/20231013-authorization-roles.md

```diff
@@ -0,0 +1,165 @@
+---
+status: {proposed}
+date: {2023-10-13}
+---
+
+# Authorization Roles
+
+## Authorization
+
+Until version `v1.10.0` Epinio has a very simple authorization mechanism. A user can be `admin` or `user`.  
+An admin can perform everything, while a user can do anything but only in the namespaces he created or where has permissions.
+
+This was pretty simple to implement but not very flexible.
+
+## Decision Drivers
+
+A new more flexible authorization was needed (https://github.com/epinio/epinio/issues/1946).
+
+Some of the requirements:
+- Be easy to understand and administer
+- Provide least privileges
+- Allow separate roles for listing, reading details, editing, and creating
+- Allow soft multi-tenancy across namespaces (e.g. user A has read write access for apps on one namespace but read only on another namespace)
+- Not require too many changes to the existing code base
+
+## Considered Options
+
+- Secret with whitelist ACL per namespace
+- Role Bindings as their own CRDs
+- ConfigMaps defined Roles with predefined Actions
+
+## Decision Outcome
+
+Chosen option 3 (__"ConfigMaps defined Roles with predefined Actions"__). It was the easier and more flexible solution. It will not require many changes in the code, and it will be easy to customize and implement.
+We could also define some basic roles that can be used.
+
+
+## Pros and Cons of the Options
+
+### Secret with whitelist ACL per namespace
+
+A secret per namespace containing the whitelist ACL could provide this without too much administrative overhead.
+
+The ACL data in the secret could look something like this
+
+```yaml
+application:
+  jwt_claim_name: rwlcd
+  jwt_claim2: l
+service:
+  jwt_claim_name: rw
+  jwt_claim2: rld
+```
+
+Where the RHS of the claim mapping is a string with a combination of letters:
+
+```
+r -- read
+w -- write/edit
+l -- list
+c -- create
+d -- delete
+a -- all
+```
+
+A user with multiple claims would get access to the union of access and an admin would have access to everything.
+
+Additionally, as long as the secret/configmap is mounted in the pod without any subpath, changing it would not require a restart of the API.
+
+#### Pros
+
+* Clear "resource" and "permission" description
+
+#### Cons
+
+* Not sure how to handle newly created namespace
+* Some operations are not strictly tight to a "resource". How to avoid the `exec` but allowing the creation of apps?
+
+### Role Bindings as their own CRDs
+
+Building custom Role Bindings as their own CRDs.
+
+#### Pros
+
+* Clear and structured implementation
+
+#### Cons
+
+* Difficult/complex to implement and evolve
+* Complex to manage
+
+### ConfigMaps defined Roles with predefined Actions
+
+To avoid changing too much the current implementation we can add "roles" to an Epinio user. At the moment the user can only have one role, but this can be extended with a list instead of a fixed value. Also the role can be namescoped adding a `::` delimiter.
```

Comment:
```suggestion
To avoid changing the current implementation too much we can add "roles" to an Epinio user. While a user can only have one role at the moment, this can be extended in the future with a list instead of a single value. Furthermore a role can be namespace-scoped by adding a `::` delimiter.
```

Reply:
Question here. Why `::` for the delimiter ? What other delimiters were considered ? (`/`, `:`, `,`, ...).

Reply:
I was looking for a character that was not "reserved", or conflicting with the Kubernetes rules (btw annotations are more flexible I think). Commas are usually used to separate completely different "set", and it was already used to separate different roles. I was looking for something that seemed to be like a URN.

While testing the OIDC authentication with Github the returned groups are formed with the `org:team` pair, so it looked ok to have a similar (but different) delimiter for `role::namespace`. Also `/` could have been fine btw, but the roles annotations would have looked "weird". Checking the yaml basically the choices would have been:

```yaml
metadata:
  annotations:
    epinio.io/roles: "admin,user::workspace,reader::foobar"
---
metadata:
  annotations:
    epinio.io/roles: "admin,user/workspace,reader/foobar"
---
metadata:
  annotations:
    epinio.io/roles: "admin,user:workspace,reader:foobar"
```

This is the outcome of the `rolesMapping` section in the dex.yaml configuration:

```yaml
    - connectorId: github
      groups:
        - id: epinio:epinio
          roles:
            - admin
        - id: epinio:ui
          roles:
            - user::workspace
        - id: Org1:TeamBlue
          roles:
            - user
            - admin::workspace
```

I'm ok with changing it to something different. Maybe the single `:` (or triple? `:::`) , or the `/` also looks nice to me.

Regarding delimiters I was not sure also about the comma. Maybe a pipe `|` could have been an option easier to read:

```yaml
metadata:
  annotations:
    epinio.io/roles: "admin|user::workspace|reader::foobar"
---
metadata:
  annotations:
    epinio.io/roles: "admin|user:workspace|reader:foobar"
---
metadata:
  annotations:
    epinio.io/roles: "admin|user/workspace|reader/foobar"
```

Reply:
I am ok with the comma for role separation. I would have used `:` or `/` for the namespace separation.

Reply:
Ok, I think I will change it to a single `:` then. :+1: